### PR TITLE
Inbound queue submission no longer fails silently

### PR DESCRIPTION
**Pull request by vgeddes**
Fix for audit issue (3).

Fixes: SNO-626

### CODE REVIEW

**Comment by yrong** [comment]
Are we gonna to implement some kind of the refund mechanism as mentioned?

> We recommend implementing an error-handling mechanism in order to return tokens from the Agent to the user if the bridging process does not succeed on the Polkadot side

File: parachain/pallets/inbound-queue/src/test.rs [hunk 1]
```diff
@@ -231,13 +271,40 @@ fn test_submit_happy_path() {
...
+		// Submit message
+		let message = Message {
+			data: OUTBOUND_QUEUE_EVENT_LOG_INVALID_DEST.into(),
+			proof: Proof {
+				block_hash: Default::default(),
+				tx_index: Default::default(),
+				data: Default::default(),
+			},
+		};
+		assert_noop!(
```
Comment (claravanstaden) [inline parachain/pallets/inbound-queue/src/test.rs:301]: This is cool, leant something today. 😄

**Review by claravanstaden — APPROVED** [review]

**Review by yrong — APPROVED** [review]
LGTM!

Though my only concern about this PR is there some irrecoverable error and in that case the bridge will be blocked and there should be some remedial action(from governance) including(not limited):

- Increase the nonce of the inbound channel to bypass that particular error message
- Refund the ERC20 token locked in Agent to the original sender

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 2]
```diff
@@ -208,39 +231,24 @@ pub mod pallet {
...
 				Err(_) => {
-					Self::deposit_event(Event::MessageReceived {
-						dest: envelope.dest,
-						nonce: envelope.nonce,
-						result: MessageDispatchResult::InvalidPayload,
-					});
-					return Ok(())
-				},
+					return Err(Error::<T>::InvalidPayload.into());
+				}
```
Comment (claravanstaden) [inline parachain/pallets/inbound-queue/src/lib.rs:239]: Nit: You could shorten this:

```
Err(_) => Err(Error::<T>::InvalidPayload.into()),
```

File: parachain/pallets/inbound-queue/src/lib.rs [hunk 3]
```diff
@@ -208,39 +231,24 @@ pub mod pallet {
...
 			};
 
-			// Attempt to convert to XCM
-			let sibling_para =
-				MultiLocation { parents: 1, interior: X1(Parachain(envelope.dest.into())) };
+			// Attempt to send XCM to a dest parachain
+			let dest = MultiLocation { parents: 1, interior: X1(Parachain(envelope.dest.into())) };
+			let (xcm_hash, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```
Comment (claravanstaden) [inline parachain/pallets/inbound-queue/src/lib.rs:244]: Do we still need to do any fee handling here (outside of this issue), using the send xcm price of delivery return parameter?

**Review by alistair-singh — APPROVED** [review]
🚀

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 3 above)
Comment (vgeddes) [reply parachain/pallets/inbound-queue/src/lib.rs:225]: Not really. I did look into it, and the fee returned by `send_xcm` (i.e the fee for sending) seems to be an optional thing that chains can decide to implement.

Ultimately, on the destination chain, some account controlled by the source chain must pay fees anyway.

File: parachain/pallets/inbound-queue/src/lib.rs (hunk 2 above)
Comment (vgeddes) [reply parachain/pallets/inbound-queue/src/lib.rs]: Thanks. Fixed in latest upload.

**Pull request merged**